### DISCUSSION TITLE
[Request for Feedback] Expiration policies and lifetime changes for PATs

### DISCUSSION THREAD

**Discussion by hpsin** [community/community#141929] [Enterprise]
We've launched the ability to restrict PAT lifetimes if the PAT is going to be used against your organization or resource. As part of this change, we've also allowed fine-grained PATs to be created without an expiration. 

The default policy for all orgs and enterprises retains the existing 366 day maximum for fine-grained PATs. 

Please check out the [changelog](https://github.blog/changelog/2024-10-18-new-pat-rotation-policies-preview-and-optional-expiration-for-fine-grained-pats/) for more details, and let us know what you think! 

#### Known issues

* **Fixed:** Editing a token was triggering a 404 page with an error message about the lifetime policy.

**Comment by mve** [comment]
I'm having some trouble creating a token without an expiration.

In the org i've unchecked "Fine-grained personal access tokens must expire".
Then on [this page](https://github.com/settings/personal-access-tokens/new) I'm trying to create a new fine-grained PAT. I've set the resource owner to the org but I don't see the option to have no expiration date. I only see 7 - 90 days and custom only goes to 1 year. The org is on the "free" plan and I am an owner of the org.

**Reply by hpsin** [reply]
Thanks! This should be fixed now, we had a feature flag stuck.

**Reply by mve** [reply]
Perfect, I see the option now! Thanks for this new feature. This was the main reason I was still using classic tokens for some use cases.

**Reply by StanislavUhimenko** [reply]
DRIUJ

**Reply by StanislavUhimenko** [reply]
https://github.githubassets.com/assets/ui/packages/memex/src/client/components/board/hooks/use-is-visible.tsx

**Comment by HikaruEgashira** [comment]
I'm trying to enable this policy in the enterprise.

After enabling this option, if unexpected API call failures are discovered and the policy is temporarily disabled, will the call failures recur?

**Reply by hpsin** [reply]
No, if you disable the policy it'll stop blocking API calls. So you can "brownout" or "scream test" long lived tokens here by setting the token, checking your streaming logs, and then deciding if you want to roll back.

**Comment by hfhbd** [comment]
We would like to disable the classic PATs for our org, but unfortunately we need read Packages (maven) support only.

Is there any option to limit a classic PAT for an org to allow only specific scopes and disable all other scopes?

**Reply by hpsin** [reply]
Ah, no, I'm afraid not. The only option today would be to run a job that checks for tokens that don't match your allowlist and revoke their access to the organization, by removing their SSO credential authorization.

**Comment by StanislavUhimenko** [comment]
import memoize from '@github/memoize'
import {getEnv} from '@github-ui/client-env'
import {IS_SERVER} from '@github-ui/ssr-utils'

function getEnabledFeaturesSet(): Set<string> {
  const features = getEnv().featureFlags
  const featuresUpperCase = features.map(feature => feature.toLowerCase())
  return new Set<string>(featuresUpperCase)
}

const featuresSet =
  IS_SERVER || process.env.NODE_ENV === 'test' || isStorybook() ? getEnabledFeaturesSet : memoize(getEnabledFeaturesSet)

export function getEnabledFeatures(): string[] {
  return Array.from(featuresSet())
}

export function isFeatureEnabled(name: string): boolean {
  return featuresSet().has(name.toLowerCase())
}

//exported to allow mocking in tests
const featureFlag = {isFeatureEnabled}

export {featureFlag}

function isStorybook() {
  try {
    return process?.env?.STORYBOOK === 'true'
  } catch {
    return false
  }
}
